skorch/tests/test_history.py

```diff
@@ -218,3 +218,99 @@ def test_history_save_load_cycle_file_path(self, history, tmpdir):
         new_history = History.from_file(str(history_f))
 
         assert history == new_history
+
+    @pytest.mark.parametrize('type_', [list, tuple])
+    def test_history_multiple_keys(self, history, type_):
+        dur_loss = history[-1, type_(['duration', 'total_loss'])]
+        # pylint: disable=unidiomatic-typecheck
+        assert type(dur_loss) is type_ and len(dur_loss) == 2
+
+        loss_loss = history[-1, 'batches', -1, type_(['loss', 'loss'])]
+        # pylint: disable=unidiomatic-typecheck
+        assert type(loss_loss) is type_ and len(loss_loss) == 2
+
+    def test_history_key_in_other_epoch(self):
+        h = History()
+        for has_valid in (True, False):
+            h.new_epoch()
+            h.new_batch()
+            h.record_batch('train_loss', 1)
+            if has_valid:
+                h.new_batch()
+                h.record_batch('valid_loss', 2)
+
+        with pytest.raises(KeyError):
+            # pylint: disable=pointless-statement
+            h[-1, 'batches', -1, 'valid_loss']
+
+    def test_history_no_epochs_index(self):
+        h = History()
+        with pytest.raises(IndexError):
+            # pylint: disable=pointless-statement
+            h[-1, 'batches']
+
+    def test_history_jagged_batches(self):
+        h = History()
+        for num_batch in (1, 2):
+            h.new_epoch()
+            for _ in range(num_batch):
+                h.new_batch()
+        # Make sure we can access this batch
+        assert h[-1, 'batches', 1] == {}
+
+    @pytest.mark.parametrize('value, check_warn', [
+        ([], False),
+        (np.array([]), True),
+    ])
+    def test_history_retrieve_empty_list(self, value, check_warn, recwarn):
+        h = History()
+        h.new_epoch()
+        h.record('foo', value)
+        h.new_batch()
+        h.record_batch('batch_foo', value)
+
+        # Make sure we can access our object
+        assert h[-1, 'foo'] is value
+        assert h[-1, 'batches', -1, 'batch_foo'] is value
+
+        # There should be no warning about comparison to an empty ndarray
+        if check_warn:
+            assert not recwarn.list
+
+    @pytest.mark.parametrize('has_epoch, epoch_slice', [
+        (False, slice(None)),
+        (True, slice(1, None)),
+    ])
+    def test_history_no_epochs_key(self, has_epoch, epoch_slice):
+        h = History()
+        if has_epoch:
+            h.new_epoch()
+        # Expect a list of zero epochs
+        assert h[epoch_slice, 'batches'] == []
+        assert h[epoch_slice, 'foo'] == []
```

Comment:
I can see the argument for why `h[epoch_slice, 'batches']` should work, but why do we expect `h[epoch_slice, 'foo']` to work when there is no key `'foo'` in history?

Same question for lines 291, 303, 304, 316.

Reply:
I had originally assumed that raising KeyError in that case was an unintended change. For example, this would work before PR 312:
```python
assert history[:, 'foo'] == [1, 2]
assert history[1:, 'foo'] == [2]
assert history[2:, 'foo'] == []  # Fails after PR 312
```

But I now realize that the behavior after PR 312 is more consistent when keys may not exist:
```python
history[:, 'bar']  # Raises KeyError
history[1:, 'bar']  # Raises KeyError
history[2:, 'bar']  # Raises KeyError after PR 312
```

This example from lr_scheduler.py helped change my mind:
https://github.com/skorch-dev/skorch/blob/1d37ef3be97d0015463a4513006c970cc86c5ebe/skorch/callbacks/lr_scheduler.py#L136-L140

Assume there are no epochs, and that `if net.history` is omitted. Returning an empty list would allow the first summation to run, even if `train_batch_count` is never present. I think robust code should explicitly handle no epochs and/or KeyError anyway, so I've reverted to raising KeyError in the case of no epochs or batches, except for `history[:, 'batches']`.